Replace RecursiveMutex m_cs_banned with Mutex, and rename it

hebasto opened the pull request:
This PR is an alternative to bitcoin/bitcoin#24092. Last two commit have been cherry-picked from the latter.

w0xlt committed:
scripted-diff: rename m_cs_banned -> m_banned_mutex

-BEGIN VERIFY SCRIPT-
s() { sed -i 's/m_cs_banned/m_banned_mutex/g' $1; }
s src/banman.cpp
s src/banman.h
-END VERIFY SCRIPT-

w0xlt committed:
refactor: replace RecursiveMutex m_banned_mutex with Mutex

hebasto committed:
refactor: Get rid of `BanMan::BannedSetIsDirty()`

hebasto committed:
refactor: Get rid of `BanMan::SetBannedSetDirty()`

w0xlt reviewed (approved):
tACK 39e101f

In addition to the refactoring, this PR:

. Adds lock acquisition in `BanMan::DumpBanlist` before calling `SweepBanned()`.

. Replaces `BanMan::SetBannedSetDirty() ` with `WITH_LOCK(m_cs_banned, m_is_dirty = false);` (write operation) or simply `m_is_dirty` (non-blocking read operation).

. `SweepBanned()` no longer locks the mutex, just checks that it is held. This function is now called with `WITH_LOCK(m_banned_mutex, SweepBanned());`.

. `BanMan::DumpBanlist` called `BanMan::SweepBanned()` twice (directly and via `BanMan::GetBanned`). Now it gets rid of  `BanMan::GetBanned` and handles `banmap_t banmap`  within the lock scope.

shaavan reviewed (approved):
Code Review ACK 39e101f4a257240f098432d6fa77b322924c0109

I prefer that the **SweepBanned** function, instead of locking `m_cs_banned` by itself, now asserts it being locked by its caller function. This prevents any need for double locking, and hence `m_cs_banned` could be safely converted from **RecursiveMutex** to **Mutex**.

The renaming of `m_cs_banned` to `m_banned_mutex` is an improvement, and I agree with it.

The conversion of `m_banned_mutex` from **RecursiveMutex** to **Mutex** is very cleanly implemented and is clutter-free.

promag reviewed (commented):
Code review ACK 39e101f4a257240f098432d6fa77b322924c0109.

There is a behavior change where `m_client_interface->BannedListChanged()` is now called with the lock held. The only place that connects to that signal is in `ClientModel` where `updateBanlist` is called asynchronously, so I'd say this change is deadlock-free.

hebasto commented:
@promag 
> There is a behavior change where `m_client_interface->BannedListChanged()` is now called with the lock held.

The same behavior is on the master branch: https://github.com/bitcoin/bitcoin/blob/d0bf9bb6a539f151ec92725d20a2b6c22cb095a5/src/banman.cpp#L160-L166
isn't it?

promag commented:
@hebasto right! Only checked that from `DumpBanlist` the lock isn't held, but from `GetBanned` it is. So this change makes it consistent 🎉

MarcoFalke commented on src/banman.cpp:
```diff
@@ -39,9 +39,11 @@ BanMan::~BanMan()
-
-    if (!BannedSetIsDirty()) return;
+    {
+        LOCK(m_cs_banned);
```
first commit: if there are any races possible, shouldn't this lock be held for the whole duration of the function?

hebasto replied:
Yes. But another dedicated mutex should be used for that. No need to hold the `m_banned_mutex` during disk operations: `m_ban_db.Write(banmap)` and `LogPrint(BCLog::NET, ...)`.

Do you prefer to add it into this PR?

MarcoFalke replied:
Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

hebasto replied:
> Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

I'm not sure if I understand your Q correctly. But if you mean that `m_ban_db.Write(banmap)` modifies `m_banned`, it is not the case because a local copy `banmap` is used.

And isn't it a suboptimal design when callers of `BanMan` public functions are waiting for a relatively and unpredictably long I/O operation?

MarcoFalke replied:
Setting the dirty flag to false can only be done after the map has been written, so I don't see how a second mutex can solve this

hebasto replied:
#### Case A -- current implementation (both master and this PR)

Nothing prevents from `m_ban_db.Write(banmap)` concurrently from different threads.

#### Case B -- `BanMan::DumpBanlist()` acquires `m_cs_banned`/`m_banned_mutex` for the whole its duration

Performance degradation is possible. E.g., caller of `BanMan::IsDiscouraged()` must wait for `BanMan::DumpBanlist()` finished in another thread.

#### Case C -- `BanMan::DumpBanlist()` acquires a dedicated `m_dump_mutex` for the whole its duration

No concurrent `m_ban_db.Write(banmap)` calls. No performance degradation.

---
> Setting the dirty flag to false can only be done after the map has been written

How about the following patch:
```diff
--- a/src/banman.cpp
+++ b/src/banman.cpp
@@ -45,11 +45,12 @@ void BanMan::DumpBanlist()
         SweepBanned();
         if (!m_is_dirty) return;
         banmap = m_banned;
+        m_is_dirty = false;
     }
 
     int64_t n_start = GetTimeMillis();
-    if (m_ban_db.Write(banmap)) {
-        WITH_LOCK(m_banned_mutex, m_is_dirty = false);
+    if (!m_ban_db.Write(banmap)) {
+        WITH_LOCK(m_banned_mutex, m_is_dirty = true);
     }
 
     LogPrint(BCLog::NET, "Flushed %d banned node addresses/subnets to disk  %dms\n", banmap.size(),
```
?

MarcoFalke replied:
> How about the following patch:

Can you explain the patch?

hebasto replied:
Currently, when thread A performs `m_ban_db.Write(banmap)` thread B can set `m_is_dirty = true` which will be overwritten by thread A just after successful finish of `m_ban_db.Write(banmap)` call. This creates inconsistency between the actual `m_banned` state and the `m_is_dirty` flag.

With the suggested patch it is guaranteed that every time we access `m_is_dirty` its value is consistent with the actual `m_banned` state.

MarcoFalke replied:
But what is the point of the dirty flag when it will effectively be always true?

hebasto replied:
It's being set to `false` in `BanMan::DumpBanlist()`, it stays `false` when `m_ban_db.Write(banmap))` is successful. And it stays `false` if other member functions do no change `m_banned`.

It allow to skip `m_ban_db.Write(banmap))` for unchanged `m_banned`.

MarcoFalke replied:
So you are suggesting this patch combined with a dedicated dump mutex?

hebasto replied:
Yes, but in another PR to keep this one focused on the transition form `RecursiveMutex` to `Mutex`.

MarcoFalke replied:
Strictly speaking, fixing the logical race is a bugfix, so it might be better to do it first to allow easier cherry-picking to older commits. Would you mind going with the fix first?

hebasto replied:
Sure!

hebasto replied:
> Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

Please check out  `DumpMempool` implementation.

hebasto replied:
> Would you mind going with the fix first?

Done in #24168.

hebasto commented:
@MarcoFalke 
> Strictly speaking, fixing the logical race is a bugfix, so it might be better to do it first to allow easier cherry-picking to older commits. Would you mind going with the fix first?

I split bugfixes into #24168.

---

Fellow reviewers, please review #24168 first.

Making this to a draft for now.

hebasto commented:
Rebased 39e101f4a257240f098432d6fa77b322924c0109 -> 0ad0288dcb462e4156ba8f619eebe2fa321ce7af ([pr24097.02](https://github.com/hebasto/bitcoin/commits/pr24097.02) -> [pr24097.03](https://github.com/hebasto/bitcoin/commits/pr24097.03)) on top of the merged #24168.

shaavan commented on src/banman.cpp:
```diff
@@ -45,16 +45,17 @@ void BanMan::DumpBanlist()
     int64_t n_start = GetTimeMillis();
     if (!m_ban_db.Write(banmap)) {
-        SetBannedSetDirty(true);
+        LOCK(m_banned_mutex);
+        m_is_dirty = true;
```
**Doubt:** Why did you prefer this way of writing code over using `WITH_LOCK`, which you discussed with @MarcoFalke in the earlier comment.

```diff
--- a/src/banman.cpp
+++ b/src/banman.cpp
@@ -45,11 +45,12 @@ void BanMan::DumpBanlist()
         SweepBanned();
         if (!m_is_dirty) return;
         banmap = m_banned;
+        m_is_dirty = false;
     }
 
     int64_t n_start = GetTimeMillis();
-    if (m_ban_db.Write(banmap)) {
-        WITH_LOCK(m_banned_mutex, m_is_dirty = false);
+    if (!m_ban_db.Write(banmap)) {
+        WITH_LOCK(m_banned_mutex, m_is_dirty = true);
     }
 
     LogPrint(BCLog::NET, "Flushed %d banned node addresses/subnets to disk  %dms\n", banmap.size(),
```

hebasto replied:
With `m_is_dirty = false;` a few lines above I think it's more readable, no?

shaavan replied:
Makes sense. So it was an aesthetic choice, which increases code readability, and both of these don't have any functional difference. Am I correct in my analysis?

hebasto replied:
Yes, you are.

shaavan replied:
Got it. Thanks, @hebasto!

shaavan commented on src/banman.cpp:
```diff
@@ -165,18 +166,19 @@ bool BanMan::Unban(const CSubNet& sub_net)
 void BanMan::SweepBanned()
 {
+    AssertLockHeld(m_banned_mutex);
+
```
nit: This extra line can be removed.

```suggestion
```

hebasto replied:
Of course, it's only a personal opinion, but for it is not an `extra` line. It is a line which separates assertions from the function logic that helps to make the presence of assertions clearer.

shaavan replied:
I agree with you. However, I think it is better to follow a code uniformity. And some places have a line between, and some don't. This breaks the code uniformity. But these are just my 2 cents, and I am okay with whatever you think would be the optimal choice in this matter.

shaavan reviewed (commented):
Concept ACK

I had a nit suggestion and a doubt that I would like to discuss.

hebasto commented:
Converted to draft for now. Please review #25149 first.

hebasto committed:
refactor: Add more negative `!m_banned_mutex` thread safety annotations

Could be verified with
$ ./configure CC=clang CXX=clang++ CXXFLAGS='-Wthread-safety -Wthread-safety-negative'
$ make clean
$ make 2>&1 | grep m_banned_mutex

theStack commented:
Concept ACK

theStack reviewed (approved):
Code-review ACK 37d150d8c5ffcb2bddcd99951a739e97571194c7

vasild commented on src/banman.h:
```diff
@@ -60,40 +60,37 @@ class BanMan
-    void Ban(const CSubNet& sub_net, int64_t ban_time_offset = 0, bool since_unix_epoch = false);
-    void Discourage(const CNetAddr& net_addr);
-    void ClearBanned();
+    void Ban(const CNetAddr& net_addr, int64_t ban_time_offset = 0, bool since_unix_epoch = false) EXCLUSIVE_LOCKS_REQUIRED(!m_banned_mutex);
+    void Ban(const CSubNet& sub_net, int64_t ban_time_offset = 0, bool since_unix_epoch = false) EXCLUSIVE_LOCKS_REQUIRED(!m_banned_mutex);
```
I find it strange that there is a requirement to not hold the mutex when calling the constructor because at the time the constructor is called the mutex does not exist yet (it is a member variable).

I checked that if I remove these annotations, then indeed `-Wthread-safety-negative` produces a warning due to the `LOCK()` inside the constructor. The `LOCK()` is not needed either because the constructor will not be called concurrently with another method (or constructor). Removing the `LOCK()` of course causes warnings that touching `m_banned` requires the mutex :( I guess this is some weird limitation in the thread safety analysis.

ajtowns replied:
The constructor is `BanMan`, the annotated functions are `Ban` ?

vasild replied:
Right! :facepalm:

vasild reviewed (approved):
ACK 37d150d8c5ffcb2bddcd99951a739e97571194c7

hebasto commented:
@maflcko Mind having a look at this PR please?

hebasto commented:
@fanquake @achow101 

Is there anything else to do here?

achow101 commented:
ACK 37d150d8c5ffcb2bddcd99951a739e97571194c7
